numpy/linalg/linalg.py

```diff
@@ -2343,3 +2343,473 @@ def _multi_dot(arrays, order, i, j):
+    ``np.contract('ab,cb->', a, b)`` are prototypical matrix matrix multiplication
+    examples.  Higher dimensional matrix matrix multiplicaitons are also considered
+    such as ``np.contract('abcd,cdef', a, b)`` and ``np.contract('abcd,cefd', a,
+    b)``.  For the former, GEMM can be called without copying data; however, the
+    latter requires a copy of the second operand.  For all matrix matrix
+    multiplication examples it is beneficial to copy the data and call GEMM;
+    however, for matrix vector multiplication it is not beneficial to do so.  For
+    example ``np.contract('abcd,cd', a, b)`` will call GEMV while
+    ``np.contract('abcd,ad', a, b)`` will call einsum as copying the first operand
+    then calling GEMV does not provide a speed up compared to calling einsum.
+
+    For three or more operands contract computes the optimal order of two and
+    one operand operations.  The `optimal` path scales like N! where N is the
+    number of terms and is found by calculating the cost of every possible path and
+    choosing the lowest cost.  This path can be more costly to compute than the
+    contraction itself for a large number of terms (~N>7).  The `opportunistic`
+    path scales like N^3 and first tries to do any matrix matrix multiplications,
+    then inner products, and finally outer products.  This path usually takes a
+    trivial amount of time to compute unless the number of terms is extremely large
+    (~N>20).  The opportunistic path typically computes the most optimal path, but
+    is not guaranteed to do so.  Both of these algorithms are sieved by the
+    variable memory to prevent very large tensors from being formed.
+
+    Examples
+    --------
+    A index transformation example, contract runs ~2000 times faster than
+    einsum even for this small example.
+
+    Note: BLAS will be True for all contractions here when everything is
+    finished.
+
+    >>> I = np.random.rand(10, 10, 10, 10)
+    >>> C = np.random.rand(10, 10)
+    >>> opt_path = contract('ea,fb,abcd,gc,hd->efgh', C, C, I, C, C, return_path=True)
+
+    >>> print(opt_path[0])
+    [(2, 0), (3, 0), (2, 0), (1, 0)]
+    >>> print(opt_path[1])
+    Complete contraction:  ea,fb,abcd,gc,hd->efgh
+           Naive scaling:   8
+    --------------------------------------------------------------------------------
+    scaling   BLAS                  current                                remaining
+    --------------------------------------------------------------------------------
+       5     False            abcd,ea->bcde                      fb,gc,hd,bcde->efgh
+       5     False            bcde,fb->cdef                         gc,hd,cdef->efgh
+       5     False            cdef,gc->defg                            hd,defg->efgh
+       5     False            defg,hd->efgh                               efgh->efgh
+    >>> ein_result = np.einsum('ea,fb,abcd,gc,hd->efgh', C, C, I, C, C, path=opt_path[0])
```

Comment:
There is no line calculating `opt_result`, and this one computing `ein_result` is passing a wrong keyword argument `pat` to `np.einsum`.